stdlib/asm/crypto/fri/frif4e2.masm

```diff
@@ -0,0 +1,324 @@
+#!
+#!  1) ptr0 is a pointer to the first layer commitment denoted throughout the code by C. ptr0 + 1 
+#!  points to the first [d_size, t_depth, alpha1, alpha0] where d_size is the size of initial domain
+#!  divided by 4, t_depth is the depth of the Merkle tree commiment to the first layer and (alpha1, alpha0)
+#!  is the first challenge used in folding the first layer.
+#!  2) ptr_rem_start is a pointer to the first tuple of two consecutive degree 2 extension field elements 
+#!  making up the remainder codeword. This codeword can be of length either 32 or 64.
+#!  3) ptr_queries_start is a pointer to a list of tuples of the form (0, p, e1, e0) where p is a query
+#!  index at the first layer and (e0, e1) is an extension field elements corresponding to the value 
+#!  of the first layer at index p.
+#!  4) ptr_queries_end is a pointer to the first empty memory address after the last (0, p, e1, e0) tuple.
+#!
+#!  The memory referenced above is used contiguously. This means for example that:
+#!  1)  ptr_rem_start - 1 points to the last (d_size, t_depth, alpha1, alpha0) tuple.
+#!  2) The length of the remainder codeword is 2 * (ptr_rem_start - ptr_queries_start).
+#! Input: [ptr_query_start, ptr_query_end, ptr0, ptr_rem_start, g]
+#! Output: ()
+#! Cost: 2626 + 8 + num_queries * (37 + num_layers * 105 + 58)
+export.verify.1
+
+    # Store [ptr_query_start, ptr_query_end, ptr0, ptr_rem_start] to keep track of all queries
+    locaddr.0
+    mem_storew
+
+    # Compute size of remainder divided by 2 to be used in determining the appropriate remainder
+    # verification procedure.
+    # [ptr_query_start, ptr_query_end, ptr0, ptr_rem_start, g, rem_len]
+    dup
+    dup.4
+    sub
+    movdn.5
+    
+    # Compute number of remaining queries to be verified and assert it is not 0
+    # [?, ptr_query_start, ptr0, ptr_rem_start, g, rem_len, ...]
+    swap
+    dup.1
+    neq
+
+    while.true
+        # Load [0, p, e1, e0] from mempry i.e. next query data
+        # [p, e1, e0, ptr0, ptr_rem_start, g, rem_len, ...]
+        push.0.0.0.0
+        movup.4
+        mem_loadw
+        drop
+
+        # Compute poe = g^p
+        # [poe, p, e1, e0, ptr0, ptr_rem_start, rem_len, ..]
+        dup.5
+        dup.1
+        exp.u32
+
+        # We now have everything to verify query p
+        exec.verify_query
+
+        # Prepare for next iteration
+        # [?, ptr_query_start + 1, ptr0, ptr_rem_start, g, rem_len, ...]
+        dropw drop drop
+        locaddr.0
+        mem_loadw
+        add.1
+        locaddr.0
+        mem_storew
+        swap
+        dup.1
+        neq
+    end
+
+    # Verify that remainder corresponds to a low degree polynomial
+
+    # Check if remainder is of size 64
+    # [?, ptr_remainder_start]
+    drop drop swap drop
+    swap
+    push.32
+    eq
+
+    if.true
+        exec.ext2fri::verify_remainder_64
+    else
+        exec.ext2fri::verify_remainder_32
+    end
```

Comment:
There might be a bug here: it seems like this will run `ext2fri::verify_remainder_64` if the remainder size is 32 and `exec.ext2fri::verify_remainder_32` otherwise - or am I misunderstanding this?

Also, in the `else` branch, should we make sure that we get the expected length?

Reply:
Yes, 32 is used because a length 64 remainder will occupy 32 contiguous memory addresses. 32 is computed as the result of the difference of the two pointers delimiting the remainder as it is laid out in memory i.e. `ptr_query_start - ptr_rem_start`.
The soundness of the else branch is guaranteed by the caller who should either lay out a length 64 or 32 remainder codeword starting at address `ptr_rem_start` and ending at `ptr_query_start - 1`.
Am I thinking correctly here?

Reply:
Ah yes - makes sense!